third_party/amd/lib/TritonAMDGPUToLLVM/OptimizeLDSUsage.cpp

```diff
@@ -0,0 +1,305 @@
+} // namespace triton
+} // namespace mlir
+
+namespace {
+
+constexpr int LDSSize = 65536;
+constexpr int kPtrBitWidth = 64;
+
+static int getCvtOpLDSUsage(triton::gpu::ConvertLayoutOp &cvtOp) {
+  unsigned inVec = 0;
+  unsigned outVec = 0;
+  auto smemShape = triton::getScratchConfigForCvtLayout(cvtOp, inVec, outVec);
+  unsigned elems =
+      std::accumulate(smemShape.begin(), smemShape.end(), 1, std::multiplies{});
+  auto srcType = cvtOp.getSrc().getType();
+  auto bytes =
+      srcType.getElementType().isa<triton::PointerType>()
+          ? elems * kPtrBitWidth / 8
+          : elems * std::max<int>(8, srcType.getElementTypeBitWidth()) / 8;
+
+  return bytes;
+}
+
+bool isPowerOfTwo(unsigned x) { return x && (x & (x - 1)) == 0; }
+
+static std::vector<SmallVector<unsigned>> factorizePowerOf2(int n) {
+  assert(isPowerOfTwo(n));
+  int x = log2(n);
+  std::vector<SmallVector<unsigned>> pairs;
+
+  for (int i = 0; i <= x / 2; ++i) {
+    int j = x - i;
+    SmallVector<unsigned> sample(2);
+    sample[0] = pow(2, i);
+    sample[1] = pow(2, j);
+    pairs.push_back(sample);
+    std::swap(sample[0], sample[1]);
+    pairs.push_back(sample);
+  }
+
+  return pairs;
+}
+
+/**
+ * @brief Copy given layout with different warpsPerCTA parameter
+ * @param layout original layout
+ * @param warpsPerCTA new warpsPerCTA
+ * @return create layout
+ */
+Attribute createTmpLayout(Attribute layout, ArrayRef<unsigned> warpsPerCTA) {
+  auto ctx = layout.getContext();
+  if (auto src = layout.dyn_cast<triton::gpu::AMDMfmaEncodingAttr>())
+    return triton::gpu::AMDMfmaEncodingAttr::get(
+        ctx, src.getVersionMajor(), src.getVersionMinor(), warpsPerCTA,
+        src.getMDim(), src.getNDim(), src.getIsTransposed(),
+        src.getCTALayout());
+  if (auto src = layout.dyn_cast<triton::gpu::AMDWmmaEncodingAttr>())
+    return triton::gpu::AMDWmmaEncodingAttr::get(ctx, warpsPerCTA,
+                                                 src.getCTALayout());
+  if (auto src = layout.dyn_cast<triton::gpu::BlockedEncodingAttr>())
+    return triton::gpu::BlockedEncodingAttr::get(
+        ctx, src.getSizePerThread(), src.getThreadsPerWarp(), warpsPerCTA,
+        src.getOrder(), src.getCTALayout());
+  if (auto src = layout.dyn_cast<triton::gpu::DotOperandEncodingAttr>()) {
+    return triton::gpu::DotOperandEncodingAttr::get(
+        ctx, src.getOpIdx(), createTmpLayout(src.getParent(), warpsPerCTA),
+        src.getKWidth());
+  }
+  if (auto src = layout.dyn_cast<triton::gpu::SliceEncodingAttr>())
+    return triton::gpu::SliceEncodingAttr::get(
+        ctx, src.getDim(), createTmpLayout(src.getParent(), warpsPerCTA));
+  assert("Encountered unsupported layout");
+  return Attribute();
+}
+
+/**
+ * Creates two chained convert layout operations
+ *
+ * %1 = cvtOp %0 (srcLayout -> dstLayout)
+ * ->
+ * %1 = cvtOp %0 (srcLayout -> dstLayout)
+ * %2 = cvtOp %0 (srcLayout -> tmpLayout)
+ * %3 = cvtOp %1 (tmpLayout -> dstLayout)
```

Comment:
Should this be `%1 = cvtOp %2`?

Reply:
This function creates two cvtOps based on a given cvtOps. Could you be more specific about which cvtOp is the new one and which is the old one in the comment?